sensors: m5611 barometer driver

niewim19 opened the pull request:
JIRA: PP-107

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes shortly -->
Implementation of `ms5611` driver.
 - [ Link to MS5611 datasheet](https://www.te.com/commerce/DocumentDelivery/DDEController?Action=showdoc&DocId=Data+Sheet%7FMS5611-01BA03%7FB3%7Fpdf%7FEnglish%7FENG_DS_MS5611-01BA03_B3.pdf%7FCAT-BLPS0036)
 - [ Link to AN520 application note](https://www.amsys-sensor.com/downloads/notes/MS5XXX-C-code-example-for-MS56xx-MS57xx-MS58xx-AMSYS-an520e.pdf) for MS56xx/MS57x/MS58xx pressure sensors

Configuration chosen:
 - largest oversampling rate of pressure measurement for lowest measurement noise
 - medium oversampling rate for temperature for fast conversion time
Final output data rate at 77Hz

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

<!--- In case of breaking change - please advice here what needs to be done in dependent projects. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

HBuczynski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,257 @@
+
+static int ms5611_hwSetup(ms5611_ctx_t *ctx)
+{
+	const uint8_t promAddr[6] = { PROM_ADDR_C1, PROM_ADDR_C2, PROM_ADDR_C3, PROM_ADDR_C4, PROM_ADDR_C5, PROM_ADDR_C6 };
```
```suggestion
	static const uint8_t promAddr[6] = { PROM_ADDR_C1, PROM_ADDR_C2, PROM_ADDR_C3, PROM_ADDR_C4, PROM_ADDR_C5, PROM_ADDR_C6 };
```

HBuczynski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,257 @@
+	usleep(RESET_SLEEP);
+
+	/* PROM reading */
+	for (i = 0; i < sizeof(promAddr); i++) {
+		cmd = CMD_READ_PROM | promAddr[i];
+		/* data received from MS5611 is written to incremented index in ctx to match datasheet (so that C1 corresponds to promC[1] and so on) */
+		if (sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &cmd, sizeof(cmd), &ctx->promC[i + 1], sizeof(ctx->promC[i + 1]), sizeof(cmd)) < 0) {
+			fprintf(stderr, "ms5611: failed read PROM C%i\n", i + 1);
+			return -1;
+		}
+		ctx->promC[i + 1] = (ctx->promC[i + 1] << 8) | (ctx->promC[i + 1] >> 8); /* swap higher and lower bytes */
+	}
```
The documentations says:
`Address 0 contains factory data and the setup, addresses 1-6 calibration coefficients and address 7 contains
the serial code and CRC.`

If you don't want to calculated CRC, change the `ctx->promC` size. Currently the array has a size 7 but you use only 6 elements. According to documentation, the size should be 8.

niewim19 replied:
Added separate commit with change to prom reading so that whole PROM memory is read.

HBuczynski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,257 @@
+		ctx->promC[i + 1] = (ctx->promC[i + 1] << 8) | (ctx->promC[i + 1] >> 8); /* swap higher and lower bytes */
+	}
+
+	printf("C1: %i C2: %i C3: %i C4: %i C5: %i C6: %i\n", ctx->promC[1], ctx->promC[2], ctx->promC[3], ctx->promC[4], ctx->promC[5], ctx->promC[6]);
```
I suppose that it is a debugging msg. It will not provide any valuable information for user in runtime.

HBuczynski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,257 @@
+
+
+/* Sends conversion request `convCmd`, waits `usDelat` for conversion and reads ADC to `res` */
+static inline int ms5611_measure(ms5611_ctx_t *ctx, uint8_t convCmd, time_t usDelay, uint32_t *res)
```
```suggestion
static int ms5611_measure(ms5611_ctx_t *ctx, uint8_t convCmd, time_t usDelay, uint32_t *res)
```

HBuczynski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,257 @@
+
+	usleep(usDelay);
+
+	if (sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &readCmd, sizeof(readCmd), &val, sizeof(val), sizeof(readCmd)) < 0) {
```
Ups, ugly copy-pasting.

```suggestion
	if (sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &readCmd, sizeof(readCmd), val, sizeof(val), sizeof(readCmd)) < 0) {
```

niewim19 replied:
Thanks for noticing it! 

It luckily worked as address of a static array has the same value as the first element of this array (yet different sizes). Changed as suggested.

HBuczynski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,257 @@
+
+	/* Translate temperature */
+	dT = (int32_t)D2 - (((int32_t)ctx->promC[5]) << 8);           /* dT = D2 - C5 * 2^8 */
+	temp = 2000 + ((int64_t)dT * ((int64_t)ctx->promC[6]) >> 23); /* TEMP = 2000 + dT * C6 / 2^23 */
```
Apply in other places.
```suggestion
	temp = 2000 + (((int64_t)dT * (int64_t)ctx->promC[6]) >> 23); /* TEMP = 2000 + dT * C6 / 2^23 */
```

niewim19 commented:
Changes made:
 - implemented reviewdog suggestions
 - added reading whole PROM memory (as separate commit)
 - added crc check of prom memory (as separate commit)

agkaminski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,280 @@
+		word = (cnt == (PROM_SIZE * 2 - 1)) ? 0 : prom[cnt >> 1];
+
+		/* High/low byte selection */
+		if (cnt & 1 == 1) {
```
As suggested by warning
```suggestion
		if ((cnt & 1) == 1) {
```

agkaminski replied:
btw, we can do `rem ^=` based on this if and skip redundant shift left on LSB case.

agkaminski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,280 @@
+	/* Operation is performed on bytes */
+	for (cnt = 0; cnt < (PROM_SIZE * 2); cnt++) {
+		/* The CRC code is calculated and written in factory with the LSB byte in the prom (prom[7]) set to 0x00 */
+		word = (cnt == (PROM_SIZE * 2 - 1)) ? 0 : prom[cnt >> 1];
```
We can skip `rem ^= `in this case instead because x XOR 0 == x

agkaminski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,280 @@
+		}
+		rem ^= word >> 8;
+
+		for (bit = 8; bit > 0; bit--) {
```
Iterator is not used in for body, we can use standard notation for a loop
```suggestion
		for (bit = 0; bit < 8; ++bit) {
```

agkaminski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,280 @@
+		rem ^= word >> 8;
+
+		for (bit = 8; bit > 0; bit--) {
+			rem = ((rem & 0x8000) > 0) ? (rem << 1) ^ 0x3000 : (rem << 1);
```
I don't know what it does, but this form looks somewhat more readable to me (just a suggestion, your choice)
```suggestion
			rem <<= 1;
			if ((rem & 0x10000) != 0) {
				rem ^= 0x3000;
			}
```

agkaminski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,280 @@
+		}
+	}
+
+	rem = (0x000f & (rem >> 12)); /* final 4-bit reminder is CRC code */
+
+	return (rem ^ 0x00);
```
x XOR 0 = x, I don't see a reason to use `0x000f` instead of `0xf`, rem is uint32_t anyway.
```suggestion
	return ((rem >> 12) & 0xf); /* final 4-bit reminder is CRC code */
```

agkaminski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,280 @@
+	}
+
+	/* If ADC is read too early then it returns 0 */
+	if (val[0] == 0 && val[1] == 0 && val[2] == 0) {
```
This check would be more efficient to perform on `res`, as this needs 3 compares instead of 1 for uint32_t.

agkaminski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,280 @@
+	int64_t off, sens;
+
+	/* Initialize temperature with too big delay for certain response */
+	while (ms5611_measure(ctx, (CMD_CVRT_TEMP | CVRT_OSR_4096), OSR_4096_SLEEP * 2, &D2) < 0)
```
These two usages of ms5611_measure (here and next usage) could result in infinite loop on HW error, not a merge blocker imho, but would be nice to implement timeout

agkaminski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,280 @@
+
+	/* breaking naming convention on purpose to match datasheet notation 1:1 */
+	uint32_t D1 = 0, D2 = 0;
+	int32_t dT;
```
From what I see it's cast to int64_t in usage anyway

niewim19 replied:
As for `dT` I agree - changed to `int64_t`.

As for `D1` and `D2` I would leave them as u32

agkaminski replied:
Yeah, I've meant only `dT`, sorry

agkaminski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,280 @@
+		if (press > 45000 && press < 120000 && temp > -4000 && temp < 8500) {
+			ctx->evtBaro.timestamp = now;
+			ctx->evtBaro.baro.pressure = press;
+			ctx->evtBaro.baro.temp = (temp / 100) + 273;
```
Nitpick - 0*C = 273,15 K
```suggestion
			ctx->evtBaro.baro.temp = (temp + 27315) / 100;
```

agkaminski replied:
I guess we don't handle rounding here properly too
```suggestion
			ctx->evtBaro.baro.temp = (temp + 27315 + 50) / 100;
```

agkaminski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,280 @@
+	/* PROM reading */
+	for (i = 0; i < PROM_SIZE; i++) {
+		cmd = CMD_READ_PROM | (i << 1); /* LSB of prom address is always 0 */
+		if (sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &cmd, sizeof(cmd), &ctx->prom[i], sizeof(ctx->prom[i]), sizeof(cmd)) < 0) {
```
Doesn't this device handle address autoincrement on register read?

niewim19 replied:
Unfortunately it does not. It is not stated in datasheet. Also it is not used in AN520 application note (there is an example code of ms5611 driver). Even if it works without any datasheet mention I would stick to loop reading.

agkaminski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,280 @@
+	int err;
+	ms5611_ctx_t *ctx = (ms5611_ctx_t *)info->ctx;
+
+	err = beginthread(ms5611_publishthr, 4, ctx->stack, sizeof(ctx->stack), info);
```
Priority of 4 seems low to me

niewim19 replied:
All sensors currently work on priority 4. It does not impose observable overhead in short term.

agkaminski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,280 @@
+	/* hardware setup of Barometer */
+	if (ms5611_hwSetup(ctx) < 0) {
+		printf("ms5611: failed to setup device\n");
+		free(ctx);
```
Perhaps we'd like to close spi here

niewim19 replied:
@HBuczynski I think this applies to all sensorhub drivers as well. Should it be corrected in all drivers now (this PR) or in separate PR?

HBuczynski replied:
This PR focuses on m5611 barometer. General changes in sensor hub should be introduced in a separate PR.

HBuczynski commented on sensors/baro/ms5611.c:
```diff
@@ -0,0 +1,300 @@
+	uint32_t rem = 0; /* crc reminder */
+
+	/* Operation is performed on bytes */
+	for (cnt = 0; cnt < (PROM_SIZE * 2); cnt++) {
```
```suggestion
	for (cnt = 0; cnt < (PROM_SIZE * sizeof(prom[0]); cnt++) {
```

HBuczynski reviewed (approved):
LGTM
